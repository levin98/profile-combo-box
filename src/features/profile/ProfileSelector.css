.dropdown:hover .dropdown-options {
    display: block;
}