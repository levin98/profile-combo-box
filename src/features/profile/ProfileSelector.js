import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEllipsis } from '@fortawesome/free-solid-svg-icons'
import { useSelector, useDispatch } from 'react-redux'
import { add, update, remove, selectProfile, setRename, setRenameText } from './profileSlice'
import './ProfileSelector.css'

const ProfileSelector = () => {
    const dispatch = useDispatch()
    const profiles = useSelector(state => state.profile.profiles)
    const selectedProfileID = useSelector(state => state.profile.selectedProfileID)
    const rename = useSelector(state => state.profile.rename)
    const renameText = useSelector(state => state.profile.renameText)

    const addProfile = () => {
        const profile = {
            id: profiles.length + 1,
            name: `Profile ${profiles.length + 1}`,
        }
        dispatch(add(profile))
    }

    const duplicateProfile = () => {
        const newProfileName = `${profiles[profiles.length-selectedProfileID].name} Duplicate 1`
        const existingProfileNames = profiles.filter(n => n.name === newProfileName || n.name.startsWith(`${newProfileName} (`))
        const profile = {
            id: profiles.length + 1,
            name: existingProfileNames.length === 0 ? newProfileName : `${newProfileName} (${existingProfileNames.length})`,
        }
        dispatch(add(profile))
    }

    const renameProfile = () => {
        dispatch(update({id: selectedProfileID, name: renameText}))
        dispatch(setRename(false))
        dispatch(setRenameText(''))
    }

    const deleteProfile = () => {
        dispatch(remove(selectedProfileID))
    }

    const onClickRename = () => {
        dispatch(setRenameText(profiles[profiles.length-selectedProfileID].name))
        dispatch(setRename(true))
    }

    return (
        <div>
            <label htmlFor="profile">Profile: </label>

            {!rename && (
                <>
                    <select name="profile" id="profile" value={selectedProfileID} onChange={e => dispatch(selectProfile(e.target.value))}>
                        {profiles.map(profile => (
                            <option key={profile.id} value={profile.id}>{profile.name}</option>
                        ))}
                    </select>

                    <div className="dropdown inline-block ml-3">
                        <button data-testid="option-button" className='bg-white px-1'><FontAwesomeIcon icon={faEllipsis} /></button>
                        <div className="dropdown-options hidden absolute bg-white">
                            <button className='px-1' onClick={addProfile}>Add</button><br />
                            <button className='px-1' onClick={onClickRename}>Rename</button><br />
                            <button className='px-1' onClick={duplicateProfile}>Duplicate</button><br />
                            <button className='px-1' onClick={deleteProfile}>Delete</button>
                        </div>
                    </div>
                </>
            )}

            {rename && (
                <>
                    <input name='profile' type="text" value={renameText} onChange={e => dispatch(setRenameText(e.target.value))} />
                    <button className='bg-white ml-2 px-1' onClick={renameProfile}>Save</button>
                </>
            )}  
            
            
        </div>
    )
}

export default ProfileSelector;