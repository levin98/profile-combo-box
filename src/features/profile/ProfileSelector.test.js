import { screen, waitForElementToBeRemoved, fireEvent } from '@testing-library/react';
import { renderWithProviders } from '../../utils/tests-utils';
import ProfileSelector from './ProfileSelector';

test('renders profile selector', () => {
  const { container } = renderWithProviders(<ProfileSelector />);
  expect(container).toMatchSnapshot();
});

test('handle add profile', async () => {
    const { container } = renderWithProviders(<ProfileSelector />);

    const addButton = screen.getByText('Add');
    addButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    const newOption = await screen.findByText('Profile 2');
    expect(newOption).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});

test('handle rename profile', async () => {
    const { container } = renderWithProviders(<ProfileSelector />);

    const renameButton = screen.getByText('Rename');
    renameButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    const input = await screen.findByRole('textbox');
    fireEvent.change(input, { target: { value: 'New Profile 1' } });

    const confirmButton = screen.getByText('Save');
    confirmButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    const newOption = await screen.findByText('New Profile 1');
    expect(newOption).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});

test('handle duplicate profile', async () => {
    const { container } = renderWithProviders(<ProfileSelector />);

    const duplicateButton = screen.getByText('Duplicate');
    duplicateButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    const newOption = await screen.findByText('Profile 1 Duplicate 1');
    expect(newOption).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});

test('handle delete profile', async () => {
    const { container } = renderWithProviders(<ProfileSelector />, {
        preloadedState: {
            profile: {
                profiles: [{ id: 1, name: 'Profile 1' }, { id: 2, name: 'Profile 2' }],
                selectedProfileID: 1,
                rename: false,
                renameText: '',
            }
        },
    });

    const deleteButton = screen.getByText('Delete');
    deleteButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));

    await waitForElementToBeRemoved(screen.queryByText('Profile 1'));
    expect(screen.queryByText('Profile 1')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
});