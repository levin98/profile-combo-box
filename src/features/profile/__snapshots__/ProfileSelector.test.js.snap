// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle add profile 1`] = `
<div>
  <div>
    <label
      for="profile"
    >
      Profile: 
    </label>
    <select
      id="profile"
      name="profile"
    >
      <option
        value="2"
      >
        Profile 2
      </option>
      <option
        value="1"
      >
        Profile 1
      </option>
    </select>
    <div
      class="dropdown inline-block ml-3"
    >
      <button
        class="bg-white px-1"
        data-testid="option-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ellipsis "
          data-icon="ellipsis"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="dropdown-options hidden absolute bg-white"
      >
        <button
          class="px-1"
        >
          Add
        </button>
        <br />
        <button
          class="px-1"
        >
          Rename
        </button>
        <br />
        <button
          class="px-1"
        >
          Duplicate
        </button>
        <br />
        <button
          class="px-1"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`handle delete profile 1`] = `
<div>
  <div>
    <label
      for="profile"
    >
      Profile: 
    </label>
    <select
      id="profile"
      name="profile"
    >
      <option
        value="2"
      >
        Profile 2
      </option>
    </select>
    <div
      class="dropdown inline-block ml-3"
    >
      <button
        class="bg-white px-1"
        data-testid="option-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ellipsis "
          data-icon="ellipsis"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="dropdown-options hidden absolute bg-white"
      >
        <button
          class="px-1"
        >
          Add
        </button>
        <br />
        <button
          class="px-1"
        >
          Rename
        </button>
        <br />
        <button
          class="px-1"
        >
          Duplicate
        </button>
        <br />
        <button
          class="px-1"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`handle duplicate profile 1`] = `
<div>
  <div>
    <label
      for="profile"
    >
      Profile: 
    </label>
    <select
      id="profile"
      name="profile"
    >
      <option
        value="2"
      >
        Profile 1 Duplicate 1
      </option>
      <option
        value="1"
      >
        Profile 1
      </option>
    </select>
    <div
      class="dropdown inline-block ml-3"
    >
      <button
        class="bg-white px-1"
        data-testid="option-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ellipsis "
          data-icon="ellipsis"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="dropdown-options hidden absolute bg-white"
      >
        <button
          class="px-1"
        >
          Add
        </button>
        <br />
        <button
          class="px-1"
        >
          Rename
        </button>
        <br />
        <button
          class="px-1"
        >
          Duplicate
        </button>
        <br />
        <button
          class="px-1"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`handle rename profile 1`] = `
<div>
  <div>
    <label
      for="profile"
    >
      Profile: 
    </label>
    <select
      id="profile"
      name="profile"
    >
      <option
        value="1"
      >
        New Profile 1
      </option>
    </select>
    <div
      class="dropdown inline-block ml-3"
    >
      <button
        class="bg-white px-1"
        data-testid="option-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ellipsis "
          data-icon="ellipsis"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="dropdown-options hidden absolute bg-white"
      >
        <button
          class="px-1"
        >
          Add
        </button>
        <br />
        <button
          class="px-1"
        >
          Rename
        </button>
        <br />
        <button
          class="px-1"
        >
          Duplicate
        </button>
        <br />
        <button
          class="px-1"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders profile selector 1`] = `
<div>
  <div>
    <label
      for="profile"
    >
      Profile: 
    </label>
    <select
      id="profile"
      name="profile"
    >
      <option
        value="1"
      >
        Profile 1
      </option>
    </select>
    <div
      class="dropdown inline-block ml-3"
    >
      <button
        class="bg-white px-1"
        data-testid="option-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ellipsis "
          data-icon="ellipsis"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="dropdown-options hidden absolute bg-white"
      >
        <button
          class="px-1"
        >
          Add
        </button>
        <br />
        <button
          class="px-1"
        >
          Rename
        </button>
        <br />
        <button
          class="px-1"
        >
          Duplicate
        </button>
        <br />
        <button
          class="px-1"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
